keras_cv/layers/regularization/dropblock_2d_test.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import tensorflow as tf
+
+from keras_cv.layers.regularization.dropblock_2d import DropBlock2D
+
+
+class DropBlock2DTest(tf.test.TestCase):
+    FEATURE_SHAPE = (1, 14, 14, 256)  # Shape of ResNet block group 3
+    rng = tf.random.Generator.from_non_deterministic_state()
+
+    def test_layer_not_created_with_invalid_block_size(self):
+        invalid_sizes = [0, -10, (5, -2), (0, 7), (1, 2, 3, 4)]
+        for size in invalid_sizes:
+            with self.assertRaises(ValueError):
+                DropBlock2D(dropblock_size=size, dropout_rate=0.1)
+
+    def test_layer_not_created_with_invalid_dropout_rate(self):
+        invalid_rates = [1.1, -0.1]
+        for rate in invalid_rates:
+            with self.assertRaises(ValueError):
+                DropBlock2D(dropout_rate=rate, dropblock_size=7)
+
+    def test_input_unchanged_in_eval_mode(self):
+        dummy_inputs = self.rng.uniform(shape=self.FEATURE_SHAPE)
+        layer = DropBlock2D(dropout_rate=0.1, dropblock_size=7)
+
+        output = layer(dummy_inputs, training=False)
+
+        tf.debugging.assert_near(dummy_inputs, output)
+
+    def test_input_unchanged_with_dropout_rate_equal_to_zero(self):
+        dummy_inputs = self.rng.uniform(shape=self.FEATURE_SHAPE)
+        layer = DropBlock2D(dropout_rate=0.0, dropblock_size=7)
+
+        output = layer(dummy_inputs, training=True)
+
+        tf.debugging.assert_near(dummy_inputs, output)
+
+    def test_input_gets_partially_zeroed_out_in_train_mode(self):
+        dummy_inputs = self.rng.uniform(shape=self.FEATURE_SHAPE)
+        layer = DropBlock2D(dropout_rate=0.1, dropblock_size=7)
+
+        output = layer(dummy_inputs, training=True)
+        num_input_zeros = self._count_zeros(dummy_inputs)
+        num_output_zeros = self._count_zeros(output)
+
+        self.assertGreater(num_output_zeros, num_input_zeros)
```

Comment:
I think we should do more strict assertion, eg we know the dropblock size is 49 (7 * 7), we should force the input to have non zeros, and count the output to have 49 zeros.

Also, we should assert the value to be properly scale up, since we know the size difference between inputs and output.

Reply:
The output is not going to have exactly 49 zeros. 

The DropBlock does not always drop exactly one, randomly placed block. It can drop 1, 2, 3 or more blocks from a single channel. It can also not drop any block at all. If multiple blocks are dropped, the dropped values can overlap.
Therefore the output (single) feature map is not going to have 7x7=49 zeros. It can have zero zeros. Even if there are 2 blocks it might not have 2x7x7 zeros as these blocks can overlap. It is impossible to tell the outcome.

Because we cannot predict the number of zeros it is also impossible to predict the scale values.

Reply:
I see. Thanks for the explanation. The example in the docstring somehow give me the impression that it will drop only one block (and for all the channels). Maybe we should provide a more generic example in the docstring?

Reply:
I agree, this docstring might have been misleading - sorry for that. Added second example of applying the layer on a deeper feature map which showcases most possible outcomes.